src/webui/api/appcontroller.cpp

```diff
@@ -1312,6 +1357,89 @@ void AppController::setCookiesAction()
     setResult(QString());
 }
 
+void AppController::clientDataAction()
+{
+    const QString keysParam {params()[u"keys"_s]};
+    if (keysParam.isEmpty())
+    {
+        setResult(m_clientData);
+        return;
+    }
+
+    QJsonParseError jsonError;
+    const auto keysJsonDocument = QJsonDocument::fromJson(keysParam.toUtf8(), &jsonError);
+    if (jsonError.error != QJsonParseError::NoError)
+        throw APIError(APIErrorType::BadParams, jsonError.errorString());
+    if (!keysJsonDocument.isArray())
+        throw APIError(APIErrorType::BadParams, tr("keys must be an array"));
+
+    QJsonObject clientData;
+    for (const QJsonValue &keysJsonVal : asConst(keysJsonDocument.array()))
+    {
+        if (!keysJsonVal.isString())
+            throw APIError(APIErrorType::BadParams, tr("key must be a string"));
+
+        const QString &key = keysJsonVal.toString();
+        if (const auto iter = m_clientData.constFind(key); iter != m_clientData.constEnd())
+            clientData.insert(key, iter.value());
+    }
+
+    setResult(clientData);
+}
+
+void AppController::setClientDataAction()
+{
+    requireParams({u"data"_s});
+    QJsonParseError jsonError;
+    const auto dataJsonDocument = QJsonDocument::fromJson(params()[u"data"_s].toUtf8(), &jsonError);
+    if (jsonError.error != QJsonParseError::NoError)
+        throw APIError(APIErrorType::BadParams, jsonError.errorString());
+    if (!dataJsonDocument.isObject())
+        throw APIError(APIErrorType::BadParams, tr("data must be an object"));
+
+    QJsonObject clientData = m_clientData;
+    bool dataModified = false;
+    const QJsonObject dataJsonObject = dataJsonDocument.object();
+    for (auto it = dataJsonObject.constBegin(), end = dataJsonObject.constEnd(); it != end; ++it)
+    {
+        const QString &key = it.key();
+        const QJsonValue &value = it.value();
+
+        if (value.isNull())
+        {
+            if (auto it = clientData.find(key); it != clientData.end())
+            {
+                clientData.erase(it);
+                dataModified = true;
+            }
+        }
+        else
+        {
+            const auto &existingValue = clientData.constFind(key);
+            if ((existingValue == clientData.constEnd()) || (existingValue.value() != value))
+            {
+                clientData.insert(key, value);
+                dataModified = true;
+            }
```

Comment:
A bit of optimization for the case where the data exists (this is supposed to be a more frequent case):
```suggestion
            const auto &existingValue = clientData.find(key);
            if (existingValue == clientData.end())
            {
                clientData.insert(key, value);
                dataModified = true;
            }
            else if (existingValue.value() != value)
            {
                existingValue.value() = value;
                dataModified = true;
            }
```